File: api/src/main/java/dev/obscuria/fragmentum/content/util/easing/EasingFunction.java
package dev.obscuria.fragmentum.content.util.easing;

import org.apache.commons.lang3.NotImplementedException;

@SuppressWarnings("unused")
@FunctionalInterface
public interface EasingFunction
{
    float compute(float delta);

    default EasingFunction reverse()
    {
        throw new NotImplementedException();
    }

    default EasingFunction scale(float scale)
    {
        throw new NotImplementedException();
    }

    default EasingFunction merge(EasingFunction other, float ratio)
    {
        throw new NotImplementedException();
    }

    default EasingFunction mergeOut(EasingFunction other, float ratio)
    {
        throw new NotImplementedException();
    }
}

File: api/src/main/kotlin/dev/obscuria/fragmentum/tools/easing/EasingFunction.kt
@file:Suppress("unused", "RedundantNullableReturnType")

package dev.obscuria.fragmentum.tools.easing

fun interface EasingFunction {

    fun compute(delta: Float): Float

    fun scale(scale: Float): EasingFunction = error("API stub")

    fun reversed(): EasingFunction = error("API stub")

    fun merge(other: EasingFunction, ratio: Float = 0.5f): EasingFunction = error("API stub")

    fun mergeOut(other: EasingFunction, ratio: Float = 0.5f): EasingFunction = error("API stub")
}